### PR TITLE
Add onRequest route hook, refactor tests and framework for route hooks

### PR DESCRIPTION
**Pull request by mcollina**
See #1587 for more details.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: docs/Hooks.md [hunk 1]
```diff
@@ -339,11 +339,22 @@ Note: using an arrow function will break the binding of this to the Fastify inst
 
 <a name="route-hooks"></a>
 ## Route level hooks
-You can declare one or more custom `preValidation` and `preHandler` hook(s) that will be unique for the route.
+You can declare one or more custom `onRequest`, `preParsing`,  `preValidation` and `preHandler` hook(s) that will be unique for the route.
```
Comment (Eomm) [inline docs/Hooks.md]:
```suggestion
You can declare one or more custom `onRequest`, `preParsing`, `preValidation`, `preHandler` and `preSerialization` hook(s) that will be unique for the route.
```

File: docs/Hooks.md [hunk 2]
```diff
@@ -358,6 +369,14 @@ fastify.route({
...
   url: '/',
   schema: { ... },
+  onRequest: function (request, reply, done) {
+    // this hook will always be executed after the shared `onRequest` hooks
+    done()
+  },
+  preParsing: function (request, reply, done) {
+    // this hook will always be executed after the shared `preParsing` hooks
+    done()
+  },
```
Comment (Eomm) [inline docs/Hooks.md]: after the `preValidation` we should add (it is missing - or another pr):

```js
  preSerialization: (request, reply, payload, next) => {
    // manipulate the payload
    next(null, payload)
  }
```

File: docs/Hooks.md [hunk 3]
```diff
@@ -339,11 +339,22 @@ Note: using an arrow function will break the binding of this to the Fastify inst
 
 <a name="route-hooks"></a>
 ## Route level hooks
-You can declare one or more custom `preValidation` and `preHandler` hook(s) that will be unique for the route.
+You can declare one or more custom `onRequest`, `preParsing`, `preValidation`, `preHandler` and `preSerialization` hook(s) that will be unique for the route.
```
Comment (jsumners) [inline docs/Hooks.md]: The implementation looks to support array values for these route hooks but the documentation does not mention such a feature.

Comment (mcollina) [reply docs/Hooks.md]: @jsumners PTAL.

**Pull request merged**